README.md
<h1 align="center">Welcome to Auto King 👋</h1>
<p>
  <img alt="Version" src="https://img.shields.io/badge/version-1.0.0-blue.svg?cacheSeconds=2592000" />
  <a href="#" target="_blank">
    <img alt="License: ISC" src="https://img.shields.io/badge/License-ISC-yellow.svg" />
  </a>
</p>

> Auto King is a punch in / out automation tool from King Of Time.

## Usage

```sh
make punch-in
make punch-out
make punch-in-schedule
make punch-out-schedule
```

## Author

👤 **locona.0810**

* Github: [@locona](https://github.com/locona)

## Show your support

Give a ⭐️ if this project helped you!

***
_This README was generated with ❤️ by [readme-md-generator](https://github.com/kefranabg/readme-md-generator)_


CHANGELOG.md
## [1.2.4](https://github.com/locona/auto-king/compare/v1.2.3...v1.2.4) (2020-09-21)


### Bug Fixes

* バグ修正 ([484586e](https://github.com/locona/auto-king/commit/484586e0fb4a42822ef9013a411f51b87bd2f441))



## [1.2.3](https://github.com/locona/auto-king/compare/v1.2.2...v1.2.3) (2020-09-08)


### Bug Fixes

* Support '有休' ([0d0a190](https://github.com/locona/auto-king/commit/0d0a190dea84db7bf527a51893da8a7f3dc6d3ea))



## [1.2.2](https://github.com/locona/auto-king/compare/v1.2.1...v1.2.2) (2020-09-02)


### Bug Fixes

* missing import utils ([081b9e9](https://github.com/locona/auto-king/commit/081b9e9ced3ef86c1200db13e80b1cd602d0b19f))



## [1.2.1](https://github.com/locona/auto-king/compare/v1.2.0...v1.2.1) (2020-09-02)


### Bug Fixes

* Bugfix deploy cmd ([96673f7](https://github.com/locona/auto-king/commit/96673f7d17e3390140b99681fba7d1a94321b08f))



# [1.2.0](https://github.com/locona/auto-king/compare/v1.1.0...v1.2.0) (2020-09-02)


### Features

* Add duplicate verification and Randomize the minutes to run ([3a19b3e](https://github.com/locona/auto-king/commit/3a19b3e2ae9f782344977914a53c526c45a8fbcc))



# 1.1.0 (2020-09-01)


### Features

* Implement Auto King ([15825ce](https://github.com/locona/auto-king/commit/15825ce55b82cb1b5120f12a9d61c13f9d442391))





functions/punchIn.js
const puppeteer = require('puppeteer');
const koyomi = require('koyomi');
const format = require('date-fns/format');
const ja = require('date-fns/locale/ja');
const utils = require('./utils')

module.exports = async (req, res) => {
  if (!utils.isRunnableMinute()) {
    res.send({'message': '出勤時間ではありません。'});
    return;
  }

  if (!utils.isWorkday()) {
    res.send({'message': '休日・祝日です。'});
    return;
  }

  const punchable = await utils.isPunchable(utils.mode.start)
  if (!punchable) {
    res.send({'message': '既に出勤済みです。'});
    return;
  }

  const browser = process.env.DEBUG
    ? await puppeteer.launch({ headless: false })
    : await puppeteer.launch({
        headless: true,
        args: ['--no-sandbox']
      });

  const page = await browser.newPage();

  const navigationPromise = page.waitForNavigation();

  await page.goto('https://s2.kingtime.jp/independent/recorder/personal/');

  await page.setViewport({ width: 1280, height: 900 });

  await navigationPromise

  await page.waitForSelector('#modal_window #id');
  await page.click('#modal_window #id');

  const user = process.env.USER;
  const password = process.env.PASSWORD;

  await page.type('#modal_window #id', user);
  await page.type('#modal_window #password', password);


  await page.waitForSelector('.modal-content > .container > .btn-control-outer > .btn-control-inner > .btn-control-message');
  await page.click('.modal-content > .container > .btn-control-outer > .btn-control-inner > .btn-control-message');

  if (process.env.DEBUG) {
    res.send({'message': 'DEBUG MODEです。'});
    return
  }

  await page.waitFor(1000);
  await page.waitForSelector('.container > #buttons > li:nth-child(1)');
  await page.click('.container > #buttons > li:nth-child(1)');

  await page.waitFor(5000);

  await browser.close();

  res.send({});
}


functions/punchOut.js
const puppeteer = require('puppeteer');
const koyomi = require('koyomi');
const format = require('date-fns/format');
const ja = require('date-fns/locale/ja');
const utils = require('./utils')

module.exports = async (req, res) => {
  if (!utils.isRunnableMinute()) {
    res.send({'message': '出勤時間ではありません。'});
    return;
  }

  if (!utils.isWorkday()) {
    res.send({'message': '休日・祝日です'});
    return;
  }

  const punchable = await utils.isPunchable(utils.mode.end)
  if (!punchable) {
    res.send({'message': '既に退勤済みです'});
    return;
  }

  const browser = process.env.DEBUG
    ? await puppeteer.launch({ headless: false })
    : await puppeteer.launch({
        headless: true,
        args: ['--no-sandbox']
      });

  const page = await browser.newPage()

  const navigationPromise = page.waitForNavigation()

  await page.goto('https://s2.kingtime.jp/independent/recorder/personal/')

  await page.setViewport({ width: 1280, height: 900 });

  await navigationPromise

  await page.waitForSelector('#modal_window #id')
  await page.click('#modal_window #id')

  const user = process.env.USER;
  const password = process.env.PASSWORD;

  await page.type('#modal_window #id', user);
  await page.type('#modal_window #password', password);

  await page.waitForSelector('.modal-content > .container > .btn-control-outer > .btn-control-inner > .btn-control-message')
  await page.click('.modal-content > .container > .btn-control-outer > .btn-control-inner > .btn-control-message')

  if (process.env.DEBUG) {
    res.send({'message': 'DEBUG MODEです。'});
    return
  }

  await page.waitFor(1000);
  await page.waitForSelector('.container > #buttons > li:nth-child(2)')
  await page.click('.container > #buttons > li:nth-child(2)')

  await page.waitFor(5000);

  await browser.close()

  res.send({});
}


functions/utils.js
const puppeteer = require('puppeteer');
const koyomi = require('koyomi');
const format = require('date-fns/format');
const ja = require('date-fns/locale/ja');

const login = async (page) => {
  const navigationPromise = page.waitForNavigation();

  await page.goto('https://s2.kingtime.jp/admin/')

  await page.setViewport({ width: 1280, height: 900 });

  await navigationPromise

  await page.waitForSelector('.specific-main #login_id')
  await page.click('.specific-main #login_id')

  const user = process.env.USER;
  const password = process.env.PASSWORD;
  await page.type('.specific-main #login_id', 'bop3003')
  await page.type('.specific-main #login_password', 'UnXqzpBaJFfqLGKhK')

  await page.waitForSelector('.specific-main #login_button')
  await page.click('.specific-main #login_button')
  await page.waitFor(2000);
}

module.exports.mode = {
  'start': 'START_TIMERECORD',
  'end': 'END_TIMERECORD'
}

module.exports.isRunnableMinute = () => {
  const today = new Date()
  const currentMonth = format(
    today,
    'MM',
    {locale: ja}
  )
  const currentDay = format(
    today,
    'dd',
    {locale: ja}
  );

  const currentMinute = format(
    today,
    'mm',
    {locale: ja}
  );


  const mon = parseInt(currentMonth, 10)
  const day = parseInt(currentDay, 10)
  const min = parseInt(currentMinute, 10)

  return day+mon <= min;
}

module.exports.isPunchable = async (mode) => {
  const today = new Date()
  const todayDay = format(
    today,
    'dd',
    {locale: ja}
  );
  const day = parseInt(todayDay, 10)

  const browser = process.env.DEBUG
    ? await puppeteer.launch({ headless: false })
    : await puppeteer.launch({
        headless: true,
        args: ['--no-sandbox']
      });
  const page = await browser.newPage();
  await login(page);

  const schedulePath = `.htBlock-adjastableTableF_inner table tbody tr:nth-child(${day}) [data-ht-sort-index='SCHEDULE']`
  page.waitForSelector(schedulePath)
  const scheduleData = await page.$eval(schedulePath, item => {
    return item.innerText;
  });

  if (scheduleData.includes('有休')) {
    return false;
  }

  const startPath = `.htBlock-adjastableTableF_inner table tbody tr:nth-child(${day}) [data-ht-sort-index='${mode}']`
  page.waitForSelector(startPath)

  const startData = await page.$eval(startPath, item => {
    return item.innerText;
  });

  await browser.close();

  if (startData) {
    return false;
  }

  return true;
}

module.exports.isWorkday = () => {
  const today = format(
    new Date(),
    'yyyy-MM-dd',
    {locale: ja}
  );
  const isOpen = koyomi.isOpen(today);
  return isOpen;
}
